classic/selection/FeatureModel.js

```diff
@@ -0,0 +1,327 @@
+                me.layer = store.getLayer();
+            }
+        }
+
+        if (!me._destroying) {
+            // bind several OL events since this is not called while destroying
+            me.bindOlEvents();
+        }
+    },
+
+    /**
+     * Binds several events on the OL objects used in this class.
+     *
+     * @private
+     */
+    bindOlEvents: function() {
+        var me = this;
+
+        // change style of selected feature
+        me.selectedFeatures.on('add', me.onSelectFeatAdd, me);
+
+        // reset style of no more selected feature
+        me.selectedFeatures.on('remove', me.onSelectFeatRemove, me);
+
+        // create a map click listener for connected vector layer
+        if (me.mapSelection && me.layer && me.map) {
+            me.map.on('singleclick', me.onFeatureClick, me);
+            me.mapClickRegistered = true;
+        }
+    },
+
+    /**
+     * Unbinds several events that were registered on the OL objects in this
+     * class (see #bindOlEvents).
+     *
+     * @private
+     */
+    unbindOlEvents: function() {
+        var me = this;
+
+        if (me.selectedFeatures) {
+            // change style of selected feature
+            me.selectedFeatures.un('add', me.onSelectFeatAdd, me);
+
+            // reset style of no more selected feature
+            me.selectedFeatures.un('remove', me.onSelectFeatRemove, me);
+        }
+
+        // create a map click listener for connected vector layer
+        if (me.mapClickRegistered) {
+            me.map.un('singleclick', me.onFeatureClick, me);
+            me.mapClickRegistered = false;
+        }
+    },
+
+    /**
+     * Handles 'add' event of #selectedFeatures.
+     * Ensures that added feature gets the #selectStyle and preserves an
+     * possibly existing feature style.
+     *
+     * @private
+     * @param  {ol.Collection.Event} evt OL event object
+     */
+    onSelectFeatAdd: function(evt) {
+        var me = this;
+        var feat = evt.element;
+        if (feat) {
+            if (feat.getStyle()) {
+                // we have to preserve the existing feature style
+                var fid = feat.getId() || me.getRandomFid();
+                me.existingFeatStyles[fid] = feat.getStyle();
+                feat.setId(fid);
+            }
+            // apply select style
+            feat.setStyle(me.selectStyle);
+        }
+    },
+
+    /**
+     * Handles 'remove' event of #selectedFeatures.
+     * Ensures that the #selectStyle is reset on the removed feature.
+     *
+     * @private
+     * @param  {ol.Collection.Event} evt OL event object
+     */
+    onSelectFeatRemove: function(evt) {
+        var me = this;
+        var feat = evt.element;
+        if (feat) {
+            var fid = feat.getId();
+            if (fid && me.existingFeatStyles[fid]) {
+                // restore existing feature style
+                feat.setStyle(me.existingFeatStyles[fid]);
```

Comment:
We should also remove the entry in the lookup, as the style might change between highlighting:

```js
delete me.existingFeatStyles[fid]
```

Reply:
Done in 4ef9166